### PR TITLE
Ensure validator indices in attester slashings are valid

### LINKED ISSUE #2345: [Fuzzing] Illegal Index Array Access in Attester Slashing Processing
### Description

While fuzzing Teku's Attester Slashing processing using [beacon-fuzz](https://github.com/sigp/beacon-fuzz), I triggered the following exception:

`Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index -1953788990 out of bounds for length 256`

Note that this was triggered using a valid `BeaconState`, and the related `AttesterSlashing` SSZ container can be successfully parsed using [zcli](https://github.com/protolambda/zcli).

From a quick glance, the affected code seems to be:

https://github.com/PegaSysEng/teku/blob/d9fdf128e70b7cc4d5d0b556f1bea20f17b7bf22/ssz/src/main/java/tech/pegasys/teku/ssz/backing/cache/ArrayIntCache.java#L68-L76

### Steps to Reproduce (Bug)

- The related `BeaconState` and `AttesterSlashing` containers can be downloaded [here](https://github.com/PegaSysEng/teku/files/4910036/teku-attester-slashing-index-array.zip).

We have a small utility to reproduce this bug, as part of `eth2fuzz`:
- Follow the instructions to set up a Teku `eth2fuzz` docker container, described [here](https://github.com/sigp/beacon-fuzz/tree/master/eth2fuzz) (i.e. `make teku`).
- Drop the two files from the archive above in `beacon-fuzz/eth2fuzz/workspace/javafuzz`
- Get a shell inside the docker container (must be ran from the `eth2fuzz` directory): 
`docker run -v $(pwd)/workspace:/eth2fuzz/workspace --entrypoint=bash -it eth2fuzz_teku`
- Move to the right directory and build the utility: `cd workspace/javafuzz/` && `javac -cp .:$(./tekuclass.sh) TekuFuzz.java`
- Run the debugging tool: `DEBUG_BEACONSTATE=beaconstate.ssz DEBUG_CONTAINER=ssz3.ssz CLASSPATH=$CLASSPATH:$(./tekuclass.sh) java TekuFuzz`
- Observe the crash:
```
[+] beaconstate ok
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index -1953788990 out of bounds for length 256
	at tech.pegasys.teku.ssz.backing.cache.ArrayIntCache.getInt(ArrayIntCache.java:70)
	at tech.pegasys.teku.ssz.backing.view.AbstractCompositeViewRead.get(AbstractCompositeViewRead.java:87)
	at tech.pegasys.teku.ssz.backing.view.AbstractCompositeViewWrite.get(AbstractCompositeViewWrite.java:86)
	at tech.pegasys.teku.ssz.backing.view.AbstractCompositeViewWrite.get(AbstractCompositeViewWrite.java:47)
	at tech.pegasys.teku.ssz.backing.view.ListViewWriteImpl.get(ListViewWriteImpl.java:95)
	at tech.pegasys.teku.ssz.backing.view.ListViewWriteImpl.get(ListViewWriteImpl.java:31)
	at tech.pegasys.teku.ssz.SSZTypes.SSZBackingList.get(SSZBackingList.java:47)
	at tech.pegasys.teku.datastructures.util.ValidatorsUtil.lambda$getValidatorPubKey$0(ValidatorsUtil.java:76)
	at tech.pegasys.teku.util.cache.LRUCache.get(LRUCache.java:73)
	at tech.pegasys.teku.datastructures.util.ValidatorsUtil.getValidatorPubKey(ValidatorsUtil.java:76)
	at tech.pegasys.teku.datastructures.util.AttestationUtil.lambda$is_valid_indexed_attestation$0(AttestationUtil.java:159)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at tech.pegasys.teku.datastructures.util.AttestationUtil.is_valid_indexed_attestation(AttestationUtil.java:159)
	at tech.pegasys.teku.datastructures.util.AttestationUtil.is_valid_indexed_attestation(AttestationUtil.java:143)
	at tech.pegasys.teku.core.operationvalidators.AttesterSlashingStateTransitionValidator.lambda$validate$1(AttesterSlashingStateTransitionValidator.java:61)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:958)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at tech.pegasys.teku.core.operationvalidators.OperationInvalidReason.firstOf(OperationInvalidReason.java:31)
	at tech.pegasys.teku.core.operationvalidators.AttesterSlashingStateTransitionValidator.validate(AttesterSlashingStateTransitionValidator.java:54)
	at tech.pegasys.teku.core.operationvalidators.AttesterSlashingStateTransitionValidator.validate(AttesterSlashingStateTransitionValidator.java:39)
	at tech.pegasys.teku.core.BlockProcessorUtil.process_attester_slashings(BlockProcessorUtil.java:318)
	at TekuFuzz.lambda$teku_attester_slashing$1(TekuFuzz.java:206)
	at tech.pegasys.teku.datastructures.state.BeaconStateImpl.updated(BeaconStateImpl.java:247)
	at TekuFuzz.teku_attester_slashing(TekuFuzz.java:204)
	at TekuFuzz.main(TekuFuzz.java:95)
```

### Versions

* Github branch: `master`
* Commit: [d9fdf12](https://github.com/PegaSysEng/teku/commit/d9fdf128e70b7cc4d5d0b556f1bea20f17b7bf22)
* Java version: `openjdk version "11.0.7" 2020-04-14`
* OS Name & Version: `Ubuntu 18.04.4 LTS`

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
`AttestaterSlashing` objects include an `IndexedAttestation` opening up the possibility that the validator indices it specifies are out of bounds for the validator list. Normally this isn't an issue because we index the attestation ourselves so know the indices are valid. 

`SignedAggregateAndProof` also includes a validator index which must be verified to be in range before looking up the validator public key.

To ensure this category of errors is avoided, `ValidatorsUtil.getValidatorPubKey` now performs bounds checking and returns an empty `Optional` if the index is out of range.  Callers are updated to handle this and fail validation of the public key is not available.

## Fixed Issue(s)
fix #2345 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

**Comment by zedt3ster** [comment]
Awesome, thanks @ajsutton.

> Normally this isn't an issue because we index the attestation ourselves so know the indices are valid.

Am I right in understanding that this bug would be triggerable when processing a malicious `AttesterSlashing` contained within a valid block?

**Comment by ajsutton** [comment]
I don't think you could create a valid block with a malicious `AttesterSlashing` - the block would be invalid if the slashing was but yes you could trigger it from within an invalid block.

**Comment by zedt3ster** [comment]
Thanks @ajsutton! Apologies, my wording in the previous comment was poor. What I meant was that nothing prevents a malicious proposer from including this `AttesterSlashing` in a proposed block. This block would fail the validation process, but in this case would cause Teku to panic. I believe we're on the same page, please feel free to correct me if I'm wrong. Thanks again for the prompt fix!

**Comment by ajsutton** [comment]
Ok yes agreed with that.  Except that panic probably isn't the right term for Teku.  It's an unhandled exception which gets logged very noisily but it doesn't cause Teku to crash.  Essentially any exception produced from processing messages (from gossip or requested blocks) just causes that message to be treated as invalid.  It's good to fix things up so that we don't log misleading error messages but the observed behaviour of Teku will wind up correct for any exception thrown from the processing or validation code.

**Comment by zedt3ster** [comment]
Great to know, thanks for clarifying!

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/AttestationUtil.java [hunk 1]
```diff
@@ -150,13 +147,17 @@ public static AttestationProcessingResult is_valid_indexed_attestation(
     SSZList<UnsignedLong> indices = indexed_attestation.getAttesting_indices();
 
     List<UnsignedLong> bit_0_indices_sorted =
-        indices.stream().sorted().distinct().collect(Collectors.toList());
+        indices.stream().sorted().distinct().collect(toList());
```
Comment (Nashatyrev) [inline ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/AttestationUtil.java:150]: Oh, is it allowed by our codestyle? I mean methods static import.
(would be just happy is yes 👍)

**Review by Nashatyrev — APPROVED** [review]
LGTM

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/AttestationUtil.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/AttestationUtil.java:150]: Yes it is as long as the method name isn't too generic (so this is fine but a static import of `valueOf` is not).  I it passes error prone and aids readability go for it. :)

**Pull request merged**